### PR TITLE
Change actor.__ray_kill__() to ray.kill(actor)

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Double-underscore methods are generally reserved for builtin keywords.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/worker.py [hunk 1]
```diff
@@ -1641,6 +1641,28 @@ def wait(object_ids, num_returns=1, timeout=None):
...
+    you can call ``actor.__ray_terminate__.remote()`` instead to queue a
+    termination task.
+
+    If this actor is reconstructable, it will be attempted to be reconstructed.
+
+    Args:
+        actor (ActorHandle): Handle to the actor to kill.
+    """
+    if not isinstance(actor, ray.actor.ActorHandle):
+        raise ValueError("ray.kill() only supported for actors.")
```
Comment (richardliaw) [inline python/ray/worker.py]: Maybe add: "Got: {type}"

Comment (edoakes) [reply python/ray/worker.py]: good idea. done

**Comment by edoakes** [comment]
cc @mitar

File: python/ray/actor.py [hunk 2]
```diff
@@ -783,18 +783,10 @@ def __del__(self):
...
-
-        Returns:
-            None.
-        """
-        worker = ray.worker.get_global_worker()
-        worker.core_worker.kill_actor(self._ray_actor_id)
+        """Deprecated - use ray.kill() instead."""
+        logger.warning("actor.__ray_kill__() is deprecated and will be removed"
+                       " in the near future. Use ray.kill(actor) instead.")
+        ray.kill(self)
```
Comment (pcmoritz) [inline python/ray/actor.py]: This won't work, right? (self is not an actor handle)

You might need to keep the old implementation here for the time being.

**Review by pcmoritz — APPROVED** [review]
LGTM but notice the comment I left

File: python/ray/actor.py (hunk 2 above)
Comment (robertnishihara) [reply python/ray/actor.py]: This is the `ActorHandle` class, so it should work, right?

File: python/ray/worker.py [hunk 3]
```diff
@@ -1641,6 +1641,29 @@ def wait(object_ids, num_returns=1, timeout=None):
...
+    """Kill an actor forcefully.
+
+    This will interrupt any running tasks on the actor, causing them to fail
+    immediately. Any atexit handlers installed in the actor will still be run.
+
+    If you want to kill the actor but let pending tasks finish,
+    you can call ``actor.__ray_terminate__.remote()`` instead to queue a
+    termination task.
+
+    If this actor is reconstructable, it will be attempted to be reconstructed.
```
Comment (robertnishihara) [inline python/ray/worker.py]: This behavior sounds right to me. Just saw this PR which might be doing the opposite (though I didn't check) https://github.com/ray-project/ray/pull/7346.

File: python/ray/worker.py [hunk 4]
```diff
@@ -1641,6 +1641,29 @@ def wait(object_ids, num_returns=1, timeout=None):
...
 
 
+def kill(actor):
+    """Kill an actor forcefully.
+
+    This will interrupt any running tasks on the actor, causing them to fail
+    immediately. Any atexit handlers installed in the actor will still be run.
+
+    If you want to kill the actor but let pending tasks finish,
+    you can call ``actor.__ray_terminate__.remote()`` instead to queue a
```
Comment (mitar) [inline python/ray/worker.py]: Egh, can we also rename `__ray_terminate__`? This is really not a good idea to use double underscores. Those are reserved for Python use.

Comment (mitar) [reply python/ray/worker.py]: Maybe the API should be `ray.terminate(actor)` and `ray.kill(actor)`. It is OK then for `ray.terminate` to call `actor.__ray_terminate__` and for subclasses of actors to subclass `__ray_terminate__` to customize its behavior. That would be aligned with Python best practices. But public API should never call `__ray_terminate__` directly.

File: python/ray/worker.py (hunk 3 above)
Comment (mitar) [reply python/ray/worker.py]: Can this be optional? Like an argument to `kill`? Or maybe `ray.restart(actor)` when you want to reconstruct it? I find it strange for this to happen automatically?

File: python/ray/actor.py (hunk 2 above)
Comment (pcmoritz) [reply python/ray/actor.py]: Ah yeah you are right

File: python/ray/worker.py (hunk 4 above)
Comment (edoakes) [reply python/ray/worker.py]: Yes, but let's discuss/do that separately.

Comment (mitar) [reply python/ray/worker.py]: I opened #7382 to followup here.

**Pull request merged**